src/rule_definition_tools.jl

```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
-            reverse_partials = [partial.args[i] for partial in partials]
-            push!(reverse_rules, rule_from_partials(inputs[i], reverse_partials...))
+
+    # For consistency in code that follows we make all partials tuple expressions
+    partials = map(partials) do partial
+        if Meta.isexpr(partial, :tuple)
+            partial
+        else
+            @assert length(inputs) == 1
+            Expr(:tuple, partial)
+        end
+    end
+
+    ############################################################################
+    # Main body: defining the results of the frule/rrule
+
+    # An expression that when evaluated will return the type of the input domain.
+    # Multiple repetitions of this expression should optimize ot. But if it does not then
+    # may need to move its definition into the body of the `rrule`/`frule`
+    𝒟 = :(typeof(first(promote($(call.args[2:end]...)))))
+
+    n_outputs = length(partials)
+    n_inputs = length(inputs)
+
+    pushforward = let
+        # Δs is the input to the propagator rule
+        # because this is push-forward there is one per input to the function
+        Δs = [Symbol(string(:Δ, i)) for i in 1:n_inputs]
+        pushforward_returns = map(1:n_outputs) do output_i
+            ∂s = partials[output_i].args
+            propergation_expr(𝒟, Δs, ∂s)
+        end
+
+        quote
+            # _ is the input derivative w.r.t. function internals. since we do not
+            # allow closures/functors with @scalar_rule, it is always ignored
+            function $(propagator_name(f, :pushforward))(_, $(Δs...))
+                return $(Expr(:tuple, pushforward_returns...))
+            end
+        end
+    end
+
+    pullback = let
+        # Δs is the input to the propagator rule
+        # because this is a pull-back there is one per output of function
+        Δs = [Symbol(string(:Δ, i)) for i in 1:n_outputs]
+
+        # 1 partial derivative per input
+        pullback_returns = map(1:n_inputs) do input_i
+            ∂s = [partial.args[input_i] for partial in partials]
+            propergation_expr(𝒟, Δs, ∂s)
+        end
+
+        quote
+            function $(propagator_name(f, :pullback))($(Δs...))
+                return (NO_FIELDS, $(pullback_returns...))
+            end
         end
-    else
-        @assert length(inputs) == 1 && all(!Meta.isexpr(partial, :tuple) for partial in partials)
-        forward_rules = Any[rule_from_partials(inputs[1], partial) for partial in partials]
-        reverse_rules = Any[rule_from_partials(inputs[1], partials...)]
     end
-    forward_rules = length(forward_rules) == 1 ? forward_rules[1] : Expr(:tuple, forward_rules...)
-    reverse_rules = length(reverse_rules) == 1 ? reverse_rules[1] : Expr(:tuple, reverse_rules...)
-    return quote
+
+    ############################################################################
+    # Final return: building the expression to insert in the place of this macro
```

Comment:
i'd still prefer this were more modular... but I really appreciate how well the code reads - good work!